### DISCUSSION TITLE
Error running input file.

### DISCUSSION THREAD

**Discussion by ahmad681** [idaholab/moose#23607] [Q&A General] [answered]
So I am using the executable to run the input file, you have the exact same one. And I keep getting this functionParser division by zero that I have no clue where it's coming from. Any idea? I am afraid this has to do with the libmesh itself. 
```
./cdd_app-opt -i inputFiles/mechanics_phase_material_interface_random.i
```

```
 nlx 18 nly 18 nlz 13
 nsx 234 nsy 234 nsz 324
 xlength 5
XMAX = 5 Y_MAX = 5 Z_MAX =  5.210260493
 MESH CREATED
 l_mesh = 0.2631578947
Setting Up
  Initializing
    Finished Initializing Equation Systems                                               [  9.43 s] [  638 MB]
  Finished Initializing                                                                  [  9.61 s] [  643 MB]
Finished Setting Up                                                                      [ 10.59 s] [  754 MB]
Framework Information:
MOOSE Version:           git commit 7ddfb7f51b on 2023-02-21
LibMesh Version:         482ca36f0ba05dbb5d87d02688d43eaa4bf983eb
PETSc Version:           3.16.6
SLEPc Version:           3.16.2
Current Time:            Sat Mar  4 18:59:38 2023
Executable Timestamp:    Sat Mar  4 18:36:19 2023
Parallelism:
  Num Processors:          1
  Num Threads:             1

Mesh:
  Parallel Type:           replicated
  Mesh Dimension:          3
  Spatial Dimension:       3
  Nodes:                   132651
  Elems:                   125000
  Num Subdomains:          1

Nonlinear System:
  Num DOFs:                397953
  Num Local DOFs:          397953
  Num Constrained DOFs:    15453
  Local Constrained DOFs:  15453
  Variables:               { "disp_x" "disp_y" "disp_z" }
  Finite Element Types:    "LAGRANGE"
  Approximation Orders:    "FIRST"

Auxiliary System:
  Num DOFs:                4512718
  Num Local DOFs:          4512718
  Variables:               { "cauchy_stress_xx" "cauchy_stress_xy" "cauchy_stress_xz" "cauchy_stress_yy"
                             "cauchy_stress_yz" "cauchy_stress_zz" } { "phase_pillar" "phase_substrate"
                             "phase_matrix" "n_pm_x" "n_pm_y" ... "n_ps_y" "n_ps_z" "n_pf_x" "n_pf_y"
                             "n_pf_z" } { "curln_pm_x" "curln_pm_y" "curln_pm_z" "divn" "elastic_strain_energy"
                             ... "interfacial_energy_pm" "interfacial_energy_ps" "interfacial_energy_ms"
                             "interfacial_energy_pf" "interfacial_energy_mf" }
  Finite Element Types:    "MONOMIAL" "LAGRANGE" "MONOMIAL"
  Approximation Orders:    "CONSTANT" "FIRST" "CONSTANT"

Execution Information:
  Executioner:             Transient
  TimeStepper:             ConstantDT
  Solver Mode:             NEWTON
  MOOSE Preconditioner:    SMP

LEGACY MODES ENABLED:
 This application uses the legacy material output option: material properties are output only on TIMESTEP_END, not INITIAL. To remove this message, set 'use_legacy_material_output' to false in this application. If there are gold output files that contain material property output for which output occurs on INITIAL, then these will generate diffs due to zero values being stored, and these tests should be re-golded.


    Finished Projecting Initial Solutions                                                [  5.30 s] [    0 MB]
ERROR: FunctionParser is unable to evaluate component 0 for 'df/dx' with arguments:
        0
        0
        -60
        0

libMesh terminating:
Reason: Division by zero
Stack frames: 22
0: libMesh::print_trace(std::ostream&)
1: libMesh::MacroFunctions::report_error(char const*, int, char const*, char const*, std::ostream&)
2: libMesh::ParsedFunction<double, libMesh::VectorValue<double> >::eval(FunctionParserADBase<double>&, std::basic_string_view<char, std::char_traits<char> >, unsigned int) const
3: MooseParsedFunctionWrapper::evaluateGradient(double, libMesh::Point const&)
4: MooseParsedFunctionTempl<Function>::gradient(double, libMesh::Point const&) const
5: ProductAux::computeValue()
6: AuxKernelTempl<double>::compute()
7: ComputeNodalAuxVarsThread<AuxKernelTempl<double> >::onNode(__gnu_cxx::__normal_iterator<libMesh::Node const* const*, std::vector<libMesh::Node const*, std::allocator<libMesh::Node const*> > >&)
8: ThreadedNodeLoop<libMesh::StoredRange<libMesh::MeshBase::const_node_iterator, libMesh::Node const*>, __gnu_cxx::__normal_iterator<libMesh::Node const* const*, std::vector<libMesh::Node const*, std::allocator<libMesh::Node const*> > > >::operator()(libMesh::StoredRange<libMesh::MeshBase::const_node_iterator, libMesh::Node const*> const&)
9: void libMesh::Threads::parallel_reduce<libMesh::StoredRange<libMesh::MeshBase::const_node_iterator, libMesh::Node const*>, ComputeNodalAuxVarsThread<AuxKernelTempl<double> > >(libMesh::StoredRange<libMesh::MeshBase::const_node_iterator, libMesh::Node const*> const&, ComputeNodalAuxVarsThread<AuxKernelTempl<double> >&)
10: void AuxiliarySystem::computeNodalVarsHelper<AuxKernelTempl<double> >(MooseObjectWarehouse<AuxKernelTempl<double> > const&, std::vector<std::vector<MooseVariableFieldBase*, std::allocator<MooseVariableFieldBase*> >, std::allocator<std::vector<MooseVariableFieldBase*, std::allocator<MooseVariableFieldBase*> > > > const&)
11: AuxiliarySystem::computeNodalVars(MooseEnumItem)
12: AuxiliarySystem::compute(MooseEnumItem)
13: FEProblemBase::computeAuxiliaryKernels(MooseEnumItem const&)
14: FEProblemBase::execute(MooseEnumItem const&)
15: FEProblemBase::initialSetup()
16: Transient::init()
17: MooseApp::executeExecutioner()
18: MooseApp::run()
19: main
20: __libc_start_main
21: ./cdd_app-opt() [0x40ef30]
[0] /home/ahmad68/projects/moose/scripts/../libmesh/installed/include/libmesh/parsed_function.h, line 719, compiled Feb 22 2023 at 13:12:24

--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_WORLD
with errorcode 1.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
```

Here if you are iterested at looking at my inputfile 

[mechanics_phase_material_interface_random.txt](https://github.com/idaholab/moose/files/10890471/mechanics_phase_material_interface_random.txt)

**Comment by GiudGiud** [comment]
Hello

I dont see anything shocking
Are you getting any deprecation warnings from this input?

What are these ProductAux for? just for a product / division?

We have a ParsedAux in the kernel that will do the same.
These ProductAux seem to be missing arguments too, I dont see what they are multiplying. 
I reckon you should transition the input file to using ParsedAux

Guillaume

**Reply by ahmad681** [reply]
Thanks for the reply Guillaume, 
so I kind of see your point, but I don't think ParsedAux does the same thing as ProductAux. 
Also I am not getting any deprecation warning. 
This used to work before just fine, maybe now Moose does things differently. I do recall that we registered ProductAux auxkernel to define the normals by taking the derivatives of a phase function. This phase function indicator was given by taking heaviside function of stepped signed function. 
So basically, we have f is the stepped signed function you can view it as a circle function for instance, sqrt((x-x0)^2-(y-y0)^2) - radius. so that would be 
f = _var1_func.value(_t, p_t)-_R[i], 
Then the phase function indicator equals 
var1+=1.0/(1+exp(-(f)/_L)); 
_L is the width of the diffuse interface basically. 
Then we apply gradient to define the normals, I hope you can see in the code I am viewing below. 

So there is this possible candidate problems I believe,  is that I traced back the error df/dx = undefined, because for some reason it is deviding by zero to find the gradient. we read locations of the pillars from an input file to initialize the problem, and now I am afraid that it is not evaluating the correct values of the pillar functions by parsing a string. What do you think?  
Another candidate problem could be maybe it is unable to evaluate the derivative at the interface when the value transition from 0 to 1, not sure to be honest. 

```
#include "ProductAux.h"
#define Y 1
#define Z 2
registerMooseObject("cddAppApp", ProductAux);

InputParameters
ProductAux::validParams()
{
	InputParameters params = AuxKernel::validParams();
	params.addParam<FunctionName>("vol1_function",0.5, "The function used to define the surface surrounding the volume 1");
	params.addParam<FunctionName>("vol2_function",0.5, "The function used to define the surface surrounding the volume 2");
	params.addRequiredParam<bool>("invert_var1","");
	params.addRequiredParam<bool>("invert_var2","");
	params.addRequiredParam<Real>("L","controls width of interface");
	params.addRequiredParam<unsigned int>("component","0=scalar representing the volume, 1=nx component, 2=ny component, 3=nz component");
	params.addParam<unsigned int>("normal_to_set",0,"The product produces 2 types of normals which join at a common line. this number sets the one the user wants to return");
	return params;
}

ProductAux::ProductAux(const InputParameters & parameters)
: AuxKernel(parameters),
  _invert1(getParam<bool>("invert_var1")),
  _invert2(getParam<bool>("invert_var2")),
  _normal_to_set(getParam<unsigned int>("normal_to_set")),
  _var1_func(getFunction("vol1_function")),
  _var2_func(getFunction("vol2_function")),
  _L(getParam<Real>("L")),
  _component(getParam<unsigned int>("component"))
{

	std::string sys_file_name = "input_pillars.txt";
	std::ifstream fileslipsys;
	MooseUtils::checkFileReadable(sys_file_name);
	fileslipsys.open(sys_file_name.c_str());
	unsigned int n_loops = 0;
	if (!(fileslipsys >> _num_pillars))
		mooseError("DensityDistributionDebugAux Error: Premature end of file reading slip system file \n");
	_centers.resize(_num_pillars);
	_R.resize(_num_pillars);
	_mag.resize(_num_pillars);
	for (unsigned int i = 0; i < _num_pillars; ++i){
		for(unsigned int j = 0; j<LIBMESH_DIM;++j){
			if (!(fileslipsys >> _centers[i](j)))
				mooseError("DensityDistributionDebugAux Error: Premature end of file reading slip system file \n");
		}
		if (!(fileslipsys >> _R[i]))
			mooseError("DensityDistributionDebugAux Error: Premature end of file reading slip system file \n");
		if (!(fileslipsys >> _mag[i]))
			mooseError("DensityDistributionDebugAux Error: Premature end of file reading slip system file \n");


	}

}

Real
ProductAux::computeValue()
{
	Real var1,var2;
	Point p;
	p(0)=(*_current_node)(0);
	p(1)=(*_current_node)(1);
	p(2)=(*_current_node)(2);
	Real f1=_var1_func.value(_t, *_current_node);
	Real f2=_var2_func.value(_t, *_current_node);
	Real surface_dirac1=1/(sqrt(M_PI)*_L)*exp( -(f1*f1)/(_L*_L) );
	Real surface_dirac2=1/(sqrt(M_PI)*_L)*exp( -(f2*f2)/(_L*_L) );
	Real tmp=0;
	var1 = 0;
	if (_invert1){//pillar
		for(unsigned int i=0;i<_num_pillars;i++){
			Point p_t = p - _centers[i];
			//Real f = sqrt(p_t(0)*p_t(0)+p_t(1)*p_t(1);
			Real f = _var1_func.value(_t, p_t)-_R[i];
			var1+=1.0/(1+exp(-(f)/_L));
		}
		var1 = var1 - (_num_pillars-1);
		var1 = 1-var1;
	}else{//matrix
		for(unsigned int i=0;i<_num_pillars;i++){
			Point p_t = p - _centers[i];
			//Real f = sqrt(p_t(0)*p_t(0)+p_t(1)*p_t(1));
			Real f = _var1_func.value(_t, p_t)-_R[i];
			var1+=1.0/(1+exp(-(f)/_L));
		}
		var1 = var1 - (_num_pillars-1);
	}
	if (_invert2){
		var2=1.0-1.0/(1.0+exp(-f2/_L));
	}else{
		var2 = 1.0/(1.0+exp(-f2/_L));
	}
	RealGradient df1 = _var1_func.gradient(_t, *_current_node);
	RealGradient df2 = _var2_func.gradient(_t, *_current_node);
	Point gradF1,gradF2;
	gradF1(0)=df1(0);gradF2(0)=df2(0);
	gradF1(1)=df1(1);gradF2(1)=df2(1);
	gradF1(2)=df1(2);gradF2(2)=df2(2);
	// var1,var2 represents heaviside(f1),heaviside(f2)
	//	Point N1 = exp(-f1/_L)/((1.0+exp(-f1/_L))*(1.0+exp(-f1/_L)))*gradF1*var2;
	Point N1;
	for(unsigned int i=0;i<_num_pillars;i++){
		Point p_t = p - _centers[i];
		RealGradient df1 = _var1_func.gradient(_t, p_t);
		gradF1(0)=df1(0);gradF1(1)=df1(1);gradF1(2)=df1(2);
		//Real f = sqrt(p_t(0)*p_t(0)+p_t(1)*p_t(1))-_R[i];
		Real f = _var1_func.value(_t, p_t)-_R[i];
		N1+=exp(-f/_L)/((1.0+exp(-f/_L))*(1.0+exp(-f/_L)))*gradF1*var2/_L;
	}

	Point N2 = exp(-f2/_L)/((1.0+exp(-f2/_L))*(1.0+exp(-f2/_L)))*gradF2*var1/_L;
	if(_component==0){
		return var1*var2;
	}else if (_normal_to_set==0){
		return N1(_component-1);
	}else{
		return N2(_component-1);
	}

}
```

**Reply by GiudGiud** [reply]
which line is doing df/dx?
how is the gradient computed? Is this from a Moose function?

**Reply by ahmad681** [reply]
To me it seems like 
```
RealGradient df1 = _var1_func.gradient(_t, *_current_node);
RealGradient df2 = _var2_func.gradient(_t, *_current_node); 
```
and 
```
RealGradient df1 = _var1_func.gradient(_t, p_t); 
```
It seems like the gradient() is being called from 
```
template <typename Output, typename OutputGradient>
inline
OutputGradient
ParsedFunction<Output,OutputGradient>::gradient (const Point & p, const Real time)
{
  OutputGradient grad;
  set_spacetime(p, time);

  grad(0) = eval(*dx_parsers[0], "df/dx", 0);
#if LIBMESH_DIM > 1
  grad(1) = eval(*dy_parsers[0], "df/dy", 0);
#endif
#if LIBMESH_DIM > 2
  grad(2) = eval(*dz_parsers[0], "df/dz", 0);
#endif

  return grad;
}
```
This can be found in 
moose>libmesh>build>include>libmesh>parsed_function.h  
I believe in my case dx_parsers[i] should be the phase function.

**Reply by GiudGiud** [reply]
Then you should have a look at the parsed expressions.

Please note that there was a rework of the parsed expression parameters, so if you are getting deprecation messages about this it might be the problem.
We renamed args -> coupled_variables
function -> expression
and a few others

**Comment by ahmad681** [comment]
Okay, the problem is very weird now. I do believe now there is nothing wrong with the code ProductAux auxillary kernel. I don't think it is even with the input file. The code works just fine but only with fewer number of elements, before it used to be with as many elements as we wanted. 
These are the new dimensions, single pillar in a matrix setting on a substrate 
    nx = 15
    ny = 15
    nz = 15
    xmin = -20
    xmax = 20
    ymin = -24.9518
    ymax = 24.9518
    zmin = -60
    zmax = 30
and the pillar is placed at (0,0,0) with radius 4. 
It is also nothing wrong with the way the parsed function is defined, I believe. I am interested to hear your thoughts about this: 
It working when I use up to 15x15x15 number of elements which is very weird. however, the shape of the pillar is not accurate because we are evaulating the function of pillar using sqrt(x^2+y^2)-radius. But when I increase the # of elements, that is when I start getting errors like 
ERROR: FunctionParser is unable to evaluate component 0 for 'df/dx' with arguments:
        0
        0
        -20
        0

libMesh terminating:
Reason: Division by zero
[0] /home/ahmad68/projects/moose/scripts/../libmesh/installed/include/libmesh/parsed_function.h, line 719, compiled Feb 22 2023 at 13:12:24

ERROR: FunctionParser is unable to evaluate component 0 for 'df/dx' with arguments:
        -20
        -24.9518
        20
        0

libMesh terminating:
Reason: Division by zero
[1] /home/ahmad68/projects/moose/scripts/../libmesh/installed/include/libmesh/parsed_function.h, line 719, compiled Feb 22 2023 at 13:12:24

ERROR: FunctionParser is unable to evaluate component 0 for 'df/dx' with arguments:
        4.44444
        -24.9518
        20
        0

libMesh terminating:
Reason: Division by zero

So there seems some problem with evaulating the gradient at (0,0,-20) center of the pillar (-20,-24.9,20) that is interface between pillar and substrate perhaps 
and (4.44,-24.9,20) is matrix/pillar/substrate common line. 

So let's see what is going on here: 
Basically I am trying to do this 
![github1](https://user-images.githubusercontent.com/77033621/223760363-8d85b901-e979-4772-a001-d7af406254e2.PNG)

And before the stress equilibrium problem was computed just fine with as many elements, so we get uniform pillar and something like this 

![github2](https://user-images.githubusercontent.com/77033621/223760413-3156be70-d494-4eb7-9933-6e54d1034cb8.PNG)

![github3](https://user-images.githubusercontent.com/77033621/223760436-57bfa6aa-14da-4a17-9316-6d718a60fba3.PNG)

Now with the new version, the code only works with 15x15x15 maximum number of elements!! so you see the pillar phase is oddly shaped in the xy plane, not along the z-direction though
![Github4](https://user-images.githubusercontent.com/77033621/223760657-11bedec1-b123-456d-8bf0-69e49bbf21e3.PNG)

![Github5](https://user-images.githubusercontent.com/77033621/223760749-9ae5128d-6b15-4e5c-a8d1-1d60cccbb473.PNG)

So what I am concluding is the following: For some reason, some points their gradient is giving undefined answer because of this devivision by zero, why is it being evaluated incorrectly there at those points, that is something that I couldn't really figure out. 
But I am happy at least that the code is functioning although with fewer number of elements. basically the larger # of elements I use, the more uniform "circular" the shape of the pillar and therefore the better results. I believe there is something with evaluating sqrt(x*x+y*y). 
Any thoughts on this, Guillaume? thanks for the help by the way

**Reply by GiudGiud** [reply]
This should be doable to debug.
You need to find the parsed function and see why it s dividing by 0 (the last component)

If you run this in a debugger, you should be able to get a backtrace
https://mooseframework.inl.gov/application_development/debugging.html

**Comment by ahmad681** [comment, marked as answer]
Final update for the discussion forum, 
The code finally works for as many elements as I want. 
 
here is what had to change: 
I  passed in to the auxillary kernels parsed functions like that: 
```
vol1_function = vol1_function_parsed
 vol2_function = vol2_function_parsed 
```
and the parsed functions are defined in Functions block like this
```
[Functions]
  [vol1_function_parsed] 
    type = ParsedFunction 
    expression = sqrt(((x-0.001)*(x-0.001))+((y-0.001)*(y-0.001)))-5  
  [../]
  [vol2_function_parsed] 
    type = ParsedFunction 
    expression = sqrt((z-20)*(z-20))-20 
  [../]
[]
```
In addition to that, the gradient of the sqrt(x^2+y^2) at the center of the circle blows up which kind of makes sense because it deviding by zero at the point, So I basically changed this to expression = sqrt(((x-0.001)*(x-0.001))+((y-0.001)*(y-0.001))) where it basically never passes the center of the pillar. 
And the solver runs pretty nicely with the correct physics.  Stresses, elastic energies, interfacial energies, strains, phase functions, etc. Thanks for the help 
![image](https://user-images.githubusercontent.com/77033621/223838053-464cc8e4-ec58-4307-8aed-00c538a98822.png)
![image](https://user-images.githubusercontent.com/77033621/223838351-4221ab7d-217c-4c8f-bb9e-0df5830c6f42.png)
